### DISCUSSION TITLE
ORM mapping registry relationship to metadata and existing tables?

### DISCUSSION THREAD

**Discussion by jlynchMicron** [sqlalchemy/sqlalchemy#10761] [Usage Questions] [answered]
Hi All,

I am creating a project where I am relying on a ORM mapping template to repetitively lay down multiple table groupings (3 relational tables per group).  As a part of this project, these table groupings may already exist on a server and I am starting to get confused between what a ORM registry represents vs a metadata object, especially around table existence (exists vs doesn't exist yet).

Can anyone explain what makes an ORM registry/mapping different from a metadata object (I do know registry's can contain a metadata object).  Also, can someone explain how the "Automap" feature may help relate these two concepts to detect if ORM objects already exist on the server or not?

Thanks!

**Comment by zzzeek** [comment]
SQLAlchemy is divided into two components, called "Core" and "ORM".  The "Core" component is completely non-dependent on ORM in any way and knows nothing about it, whereas ORM is entirely constructed around Core.

This pertains to the issue of the "MetaData" collection in that "MetaData" is a Core-concept that represents a collection of Table objects.    At this level, the ORM is not involved.     

The MetaData/  Table components are themselves datastructures exist in a Python application, and by themselves, are not related to a database server.   You can open up a Python script, make a MetaData and Table objects and now you have some datastructures, nothing happens on any kind of database.   It's basically like you drew a picture of what some database tables would look like.

The connection then between MetaData/Table and "existence" of those tables on some database server has to do with whether or not that database server has had CREATE TABLE statements emitted to it, which would establish those tables as existing.  

In SQLAlchemy, if you have a MetaData collection with some Tables in it, you can send these CREATE TABLE commands using a command called ``MetaData.create_all()``, which is passed an `Engine` object that communicates to a database server to emit these commands, effectively synchronizing that database server to have server-side table constructs that look like the MetaData/Table objects we have in our program.

After all that, the ORM wants to take Python classes and "map" them to Table objects.   These classes have their own registry which is the sqlalchemy.orm.Registry you mention.   This registry does not have any direct link to a particular MetaData collection and the two collections are independent of each other.

not sure if this helps clarify the roles of things.

**Reply by jlynchMicron** [reply]
Hi @zzzeek ,

The sqlalchemy.orm.Registry does contain its own copy of a metadata object though, correct (https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/orm/decl_api.py#L1209)?  I guess one of my main questions is, when using the ORM to a fully/semi existing database (based on the ORM), should I be using the ORM's copy of the metadata or make another metadata dedicated to the engine/session connection to the SQL server?

Thanks!

**Reply by jlynchMicron** [reply]
Part of the reason why im confused is due to the nature of my specific situation because I have 1 ORM layout, 2 SQL service (GCP + SQLite) engines; and im wondering if I should have 3 metadatas or 1.

Either:
- registery.metadata
- GCP metadata
- SQLite metadata

Or Just:
- registery.metadata

What would be the more "correct" way to set this up?

**Reply by zzzeek** [reply, marked as answer]
that's not a "copy" of the MetaData collection, that's a reference to it.   there's only one MetaData collection if you just made a registry and didn't make other MetaData collections somewhere else.

The ORM builds on top of Core.   so we have this analogy:

Core:

Metadata = collection of Tables
Table = represents a Table, used to represent something that can be SELECTed, insert/update/delete

ORM:

registry = collection of mapped classes, links to a MetaData collection of Tables
class = links to a Table, used to represent something that can be SELECTed, insert/update/delete


> Part of the reason why im confused is due to the nature of my specific situation because I have 1 ORM layout, 2 SQL service (GCP + SQLite) engines; and im wondering if I should have 3 metadatas or 1.

It generally does not really matter, the partitioning of MetaData collections you might want to do so that if you wanted to say `metadata.create_all(some_database)`, you'd only get a subset of tables that apply to that database.     Also when you create a `ForeignKey()` constraint, if you want to refer to the referred table /column using a string name, that would make use of the same `MetaData` collection and would not work as a string if the referenced table were in a different metadata.

besides those two things you could use a separate MetaData for every table, or one MetaData for everything, doesn't really matter.   However, if you have three logical databases that you tend to refer to separately when doing operations across a whole database, using three separate MetaData collections would be a common choice.

**Reply by jlynchMicron** [reply]
Thank you for the explanation!
